src/walk-manifest.js

```diff
@@ -49,116 +58,192 @@ var parseManifest = function(content) {
   return parser.manifest;
 };
 
-var parseKey = function(basedir, decrypt, resources, manifest, parent) {
-	if (!manifest.parsed.segments[0] || !manifest.parsed.segments[0].key) {
-		return {};
-	}
-	var key = manifest.parsed.segments[0].key;
-
-	var keyUri = key.uri;
-	if (!isAbsolute(keyUri)) {
-		keyUri = joinURI(path.dirname(manifest.uri), keyUri);
-	}
-
-	// if we are not decrypting then we just download the key
-	if (!decrypt) {
-		// put keys in parent-dir/key-name.key
-		key.file = basedir;
-		if (parent) {
-			key.file = path.dirname(parent.file);
-		}
-		key.file = path.join(key.file, fsSanitize(path.basename(key.uri)));
-
-		manifest.content = new Buffer(manifest.content.toString().replace(
-			key.uri,
-			path.relative(path.dirname(manifest.file), key.file)
-		));
-		key.uri = keyUri;
-		resources.push(key);
-		return key;
-	}
-
-	// get the aes key
-	var keyContent = syncRequest('GET', keyUri).getBody();
-	key.bytes = new Uint32Array([
-		keyContent.readUInt32BE(0),
-		keyContent.readUInt32BE(4),
-		keyContent.readUInt32BE(8),
-		keyContent.readUInt32BE(12)
-	]);
-
-	// remove the key from the manifest
-	manifest.content = new Buffer(manifest.content.toString().replace(
-		new RegExp('.*' + key.uri + '.*'),
-		''
-	));
-
-
-	return key;
+var parseKey = function(basedir, decrypt, resources, manifest, parent, callback) {
+  if (!manifest.parsed.segments[0] || !manifest.parsed.segments[0].key) {
+    return callback(null, {});
+  }
+  var key = manifest.parsed.segments[0].key;
+
+  var keyUri = key.uri;
+  if (!isAbsolute(keyUri)) {
+    keyUri = joinURI(path.dirname(manifest.uri), keyUri);
+  }
+
+  // if we are not decrypting then we just download the key
+  if (!decrypt) {
+    // put keys in parent-dir/key-name.key
+    key.file = basedir;
+    if (parent) {
+      key.file = path.dirname(parent.file);
+    }
+    key.file = path.join(key.file, fsSanitize(path.basename(key.uri)));
+
+    manifest.content = new Buffer(manifest.content.toString().replace(
+      key.uri,
+      path.relative(path.dirname(manifest.file), key.file)
+    ));
+    key.uri = keyUri;
+    resources.push(key);
+    return callback(null, key);
+  }
+
+  // get the aes key
+  request({url: keyUri, encoding: null, timeout: 1500}, function(error, response, body) {
+    if (error) {
+      const keyError = new Error(error.message + '|' + keyUri);
+      console.error(keyError, error);
+      return callback(keyError, {});
+    }
+    if (response.statusCode !== 200) {
+      const keyError = new Error(response.statusCode + '|' + keyUri);
```

Comment:
Should we add the response body too?

Reply:
I'd think the entire response could be passed as part of the callback error and maybe logged at a lower log level (if that makes sense with logging system). Special handling may also make sense fro 404s. I'd also be interested in what the behavior around redirects is (presumably the client will follow them so handling them shouldn't be expected).

Reply:
We could pass the response but that seems like a lot to pass to the callback. That could be a lot of object/info passed. It seems like it's enough if we see an error we just make a simple error object and then you can debug later if you want maybe - just try to walk as best effort and if some url or manifest is bad that's ok since it will get expired out the cache soon enough. I think all the status codes will be just visual checks - if you get a non 200 that means the client won't be able to play it back so we don't really care. We only care about the 200s since we really only need to purge those. Redirects are just handled by request they will follow (up to a point).

Reply:
The data passed to the callback would just be a reference to an object that already exists. Passing the full object allows the client code to decide how it wants to handle the errors which allows this library to be slightly more generalized rather than being biased towards specific consumers. Non-200s are not necessarily hard failures: from what I remember one of the major causes of issues in similar processing I've done in the past was access issues revolving around things like not including a proper token param for CDN access which is something that some calling code may be capable of and want to handle. It seems like providing a tailored projection of failure data could be helpful but dropping the rest of the data seems to needlessly limit the ways the library could be used.